query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/QueryRequestToPinotSQLConverter.java

```diff
@@ -262,22 +282,106 @@ private String convertOperator2String(Operator operator) {
     }
   }
 
+  private LiteralConstant[] getKeyValuePairForMapAttribute(Expression lhs, Expression rhs) {
+    String pathExpression = lhs.getAttributeExpression().getSubpath();
+    LiteralConstant[] kvp = convertExpressionToMapLiterals(rhs);
+    kvp[0] =
```

Comment:
Checking on the first one.

Reply:
Arrays were used previously also, so continued using them. Is there a problem using them ?

Reply:
Regarding first point, one way is to pass both lhs and rhs in the function and overload it. Should I do that ? Also all this would go away once we deprecate the use of columnIdentifier and the code would eventually become a bit simple.

Reply:
Discussed the arrays more in https://github.com/hypertrace/query-service/pull/107#discussion_r758872212 - Collections are better than arrays except for in very limited circumstances as they have a much more friendly api, both for the author and the reader. The case here calls for neither, we're using them as a shortcut to save writing a new class and that should be fixed.

> Regarding first point, one way is to pass both lhs and rhs in the function and overload it. Should I do that ?

Will hold this discussion pending the main one, but I think that likely it makes sense to pass in both yes. The question to ask is - would there ever be a case where we _don't_ want to pass in both.

>  Also all this would go away once we deprecate the use of columnIdentifier and the code would eventually become a bit simple.

If that's the case, preprocessing the query and replacing all column identifiers with their equivalent attribute expressions might be worthwhile - I suspect it's both less code than we have, and certainly less complexity. Once column identifier is fully removed, it's just one block of code to simply remove.

Reply:
resolved

Reply:
resolved